File: HW5/MISTA/CaseStudies/BlocksCaseStudy/Cpp/Blocks/Blocks/block.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

#include "Block.hpp"
using namespace std;


//**************************************************************START ON

	
	ON::ON(string a, string b)
	{
		this->a = a;
		this->b = b;
	
	}
	
	bool ON::Equals(ON on2)
	{
		return a == on2.a &&  b == on2.b;
	}

	string ON::ToString()
	{
		return "on("+a+", "+b+")";
	}

// *******************************************************************END ON


	void Block::pickup(string a)
	{
        vector<string>::iterator p;
        vector<string>::iterator p2;
        bool d, d2;
        p = find(ontables.begin(), ontables.end(), a);
        p2 = find(clears.begin(), clears.end(), a);
        if( p == ontables.end() ) {
            d = false;
        }
        else {
             d = true;
        }
        if( p2 == clears.end() ) {
            d2 = false;
        }
        else {
             d2 = true;
        }

        
		if (d && d2 && handempty)
		{
			p = ontables.erase( p );
            p2 = clears.erase( p2 );

			handempty = false;
			holding = a;
		}
		else
			cout << "Cannot pickup " << a;
	}

	void Block::putdown(string a)
	{
		if (holding!="")
		{
			holding = "";
			ontables.push_back(a);
			clears.push_back(a);
			handempty = true;
		}
		else
			cout << "Cannot put down " << a;
	}

	void Block::stack(string a, string b)
	{
        vector<string>::iterator p;
        bool d;
        p = find(clears.begin(), clears.end(), b);
        if( p == clears.end() ) {
            d = false;
        }
        else {
             d = true;
        }
 
		if (a!=b && holding!="" && holding == a && d)
		{
			handempty=true;
			ons.push_back(ON(a , b));
			p = clears.erase( p );
			clears.push_back(a);
			holding = "";
		}
		else
			cout << "Cannot stack " << a << " on " << b;
	}

	void Block::unstack(string a, string b)
	{
        vector<string>::iterator p;
        bool d;
        p = find(clears.begin(), clears.end(), a);
        if( p == clears.end() ) {
            d = false;
        }
        else {
             d = true;
        }
        
		if (handempty && d && (isOn(a, b)))
		{
			handempty = false;
			p = clears.erase( p );
			removeOn(a, b);
			holding = a;
			clears.push_back(b);
		}
		else
			cout << "Cannot unstack " << a << " from " << b;
	}

	vector<string> * Block:: getOntables()
	{
		return &ontables;
	}

	bool Block::isOntable(string a)
	{
        vector<string>::iterator p;
        bool d;
        p = find(ontables.begin(), ontables.end(), a);
        if( p == ontables.end() ) {
            d = false;
        }
        else {
             d = true;
        }
        
		return d;
	} 

	vector<string> * Block:: getClears()
	{
		return &clears;
	}

	bool Block::isClear(string a)
	{
        vector<string>::iterator p;
        bool d;
        p = find(clears.begin(), clears.end(), a);
        if( p == clears.end() ) {
            d = false;
        }
        else {
             d = true;
        }
        
		return d;
	} 

	vector<ON> * Block::getOns()
	{
		return &ons;
	}

	bool Block::isOn(string a, string b)
	{
		ON on2(a, b);
		for (unsigned on = 0; on < ons.size(); on++)
        	if (ons[on].Equals(on2))
				return true;
		return false;
	}

	void Block::removeOn(string a, string b)
	{
         vector<ON>::iterator p;
        
         ON on2(a, b);
		for (unsigned int i=ons.size() - 1; i>=0; i--)
		{            
			if (ons[i].Equals(on2))
			{
               //p = ons[i];
               //p = find(ons.begin(), ons.end(), on2);
               ons.erase(ons.begin()+i);
               // ons[i].erase;
			   //ons.remove(i);
				return;
			}
		}
	}

	string Block::getHolding()
	{
		return holding;
	}

	bool Block::isHolding(string thing)
	{
		return holding != "" && holding == thing;
	}

	bool Block::isHandempty()
	{
		return handempty;
	}

	void Block::setHandempty(bool empty)
	{
		handempty = empty;
	}

	string Block::ToString()
	{
		string str = "";
		unsigned i;
		for(i = 0; i < ontables.size(); i++)
        	str += "\n\tontable("+ontables[i]+")";
		for(i = 0; i < clears.size(); i++)
        	str += "\n\tclear("+clears[i]+")";
		for(i = 0; i < ons.size(); i++)
        	str += "\n\t" +ons[i].a+ " " +ons[i].b;
		if (holding!="")
			str += "\n\tholding("+ holding+")";
		if (handempty)
			str += "\n\thandempty(\"r\")";
		return str;
	}

File: HW5/MISTA/CaseStudies/BlocksCaseStudy/VB/Blocks/Blocks/Block.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Collections

Namespace Blocks

    Class Block

        Private ontables As New List(Of String)()
        Private clears As New List(Of String)()
        Private ons As New List(Of [OnBlock])()
        Private holding As String = Nothing
        Private handempty As Boolean = True

        Public Sub New()
        End Sub

        Public Sub Pickup(ByVal a As [String])
            If ontables.Contains(a) AndAlso clears.Contains(a) AndAlso handempty Then
                ontables.Remove(a)
                clears.Remove(a)
                handempty = False
                holding = a
            Else
                Console.WriteLine("Cannot pickup " & a)
            End If
        End Sub

        Public Sub Putdown(ByVal a As [String])
            If holding IsNot Nothing Then
                holding = Nothing
                ontables.Add(a)
                clears.Add(a)
                handempty = True
            Else
                Console.WriteLine("Cannot put down " & a)
            End If
        End Sub

        Public Sub Stack(ByVal a As [String], ByVal b As [String])
            If a <> b AndAlso holding IsNot Nothing AndAlso holding = a AndAlso clears.Contains(b) Then
                handempty = True
                ons.Add(New [OnBlock](a, b))
                clears.Remove(b)
                clears.Add(a)
                holding = Nothing
            Else
                Console.WriteLine(("Cannot stack " & a & " on ") + b)
            End If
        End Sub

        Public Sub Unstack(ByVal a As [String], ByVal b As [String])
            If handempty AndAlso clears.Contains(a) AndAlso IsOn(a, b) Then
                handempty = False
                clears.Remove(a)
                RemoveOn(a, b)
                holding = a
                clears.Add(b)
            Else
                Console.WriteLine(("Cannot unstack " & a & " from ") + b)
            End If
        End Sub

        Public Function GetOntables() As List(Of String)
            Return (ontables)
        End Function

        Public Function IsOntable(ByVal a As String) As Boolean
            Return (ontables.Contains(a))
        End Function

        Public Function GetClears() As List(Of String)
            Return (clears)
        End Function

        Public Function IsClear(ByVal a As [String]) As Boolean
            Return (clears.Contains(a))
        End Function

        Public Function GetOns() As List(Of [OnBlock])
            Return (ons)
        End Function

        Public Function IsOn(ByVal a As String, ByVal b As String) As Boolean
            Dim on2 As New [OnBlock](a, b)
            For i As Integer = 0 To ons.Count - 1
                If ons(i).Equals(on2) Then
                    Return (True)
                End If
            Next
            Return (False)
        End Function

        Private Sub RemoveOn(ByVal a As String, ByVal b As String)
            Dim on2 As New [OnBlock](a, b)
            For i As Integer = ons.Count - 1 To 0 Step -1
                If ons(i).Equals(on2) Then
                    ons.RemoveAt(i)
                    Exit Sub
                End If
            Next
        End Sub

        Public Function GetHolding() As String
            Return (holding)
        End Function

        Public Function IsHolding(ByVal thing As String) As Boolean
            Return (holding IsNot Nothing AndAlso holding.Equals(thing))
        End Function

        Public Function IsHandempty() As Boolean
            Return (handempty)
        End Function

        Public Sub SetHandempty(ByVal empty As Boolean)
            handempty = empty
        End Sub

        Public Overloads Overrides Function ToString() As [String]
            Dim str As [String] = ""
            For i As Integer = 0 To ontables.Count - 1
                str += vbLf & vbTab & "ontable(" & ontables(i) & ")"
            Next
            For i As Integer = 0 To clears.Count - 1
                str += vbLf & vbTab & "clear(" & clears(i) & ")"
            Next
            For i As Integer = 0 To ons.Count - 1
                str += vbLf & vbTab & ons(i).ToString()
            Next
            If holding IsNot Nothing Then
                str += vbLf & vbTab & "holding(" & holding & ")"
            End If
            If handempty Then
                str += vbLf & vbTab & "handempty(""r"")"
            End If
            Return (str)
        End Function

        Private Sub Init()
            ontables.Add("3")
            ontables.Add("6")
            clears.Add("1")
            clears.Add("6")
            ons.Add(New [OnBlock]("1", "3"))
            handempty = True
        End Sub

        '        Shared Sub Main()
        '        Dim block As New Block()
        '            block.Init()
        '           block.Unstack("1", "3")
        '           block.Putdown("1")
        '           block.Pickup("6")
        '           block.Stack("6", "1")
        '           Console.WriteLine("Final state: " & block.ToString())
        '       End Sub

    End Class

    Class OnBlock
        Private a As String, b As String

        Public Sub New(ByVal a As String, ByVal b As String)
            Me.a = a
            Me.b = b
        End Sub

        Public Function Equals(ByVal on2 As [OnBlock]) As Boolean
            Return (a.Equals(on2.a) AndAlso b.Equals(on2.b))
        End Function

        Public Overloads Overrides Function ToString() As String
            Return ("on(" & a & ", ") + b & ")"
        End Function

    End Class

End Namespace

